README
What is it?
===========

It is a software for the (semiautomatic) analysis and filtering of
force curves. Force curves are the output of an analytical technique
called force spectroscopy. A force spectroscopy experiment usually
requires the analysis of thousands of force curves at a time. As of
today, there is no standard, free software for the analysis of force
curves. Hooke aims to solve that.


What it does?
=============

  * View, annotate, measure force curves
  * Worm-like chain and freely-jointed chain fit of force peaks
  * Automatic convolution-based filtering of empty curves
  * Automatic fit and measurement of multiple force peaks
  * Handles force-clamp force experiments (experimental)
  * It is extensible by users by mean of plugins and drivers

See the Google Code website for more details
  http://code.google.com/p/hooke/


How do you make it work?
========================

See the doc/tutorial.txt file distributed with hooke, or the online
docs:
  http://www.physics.drexel.edu/~wking/rsrch/hooke/


Is this published in some peer-reviewed journal?
================================================

Sandal M, Benedetti F, Brucale M, Gomez-Casado A, Samorì B.
"Hooke: an open software platform for force spectroscopy."
Bioinformatics, 2009 25(11):1428-1430.
doi:10.1093/bioinformatics/btp180

Please cite Hooke if you use it.


Troubleshooting
===============

If you have troubles in using it, before throwing it in the trash:

1. Search the issue tracker.
     http://code.google.com/p/hooke/issues/list
2. Look at the trouble shooting wiki page.
     http://code.google.com/p/hooke/wiki/TroubleShooting
3. Search the discussion group.
     http://groups.google.com/group/hookesoftware
4. Ask a question in the discussion group.
5. File a bug on the issue tracker.


Disclaimer
==========

Remember that Hooke is still experimental software! It has been mostly
done to fit the needs of its authors, and there is no guarantee it
will do what you need. However, you can write us/help us improve it so
that it does. We aim to make of Hooke a robust tool that can help your
research.


CHANGELOG
THIS CHANGELOG FILE CONTANS PRE-SVN CHANGELOGS AND SUMMARIES OF POST-SVN CHANGELOGS.
FROM 0.8.4 ONWARD, DETAILED CHANGELOGS ARE AVAILABLE ON THE HOOKE SUBVERSION REPOSITORY.

0.8.4
(2008-x-x)
    PLUGINS:
    macro.py:
        hooke does not crash if it doesn't have permissions to create the folder
	fixed overwriting of export for curves with the same name and different numerical extension
	fixed macrodir (out of try-catch)
	cleaned debug output in execmacro
    DRIVERS:
	alternative version of picoforce driver
    INPUT
	merged (partially) libinput (so far in hooke_cli)
    OUTLET	
	merged liboutlet


0.8.3
(2008-04-16)
    PLUGINS:
    generalvclamp.py:
        fixed autopeak header
        fixed autopeak slope (now unwanted slope values are discarded)

0.8.2
(2008-04-10)
    PLUGINS:
    flatfilts.py:
        convfilt does not crash if a file is not a curve
    generalvclamp.py:
        autopeak now saves curve data correctly
        autopeak now generates a dummy note (so that copylog/notelog is aware you measured the curve)

0.8.1
(2008-04-07)
    PLUGINS:
    generalvclamp.py:
        fixed DeprecationWarning in flatten
    flatfilts.py
        convfilt now working


0.8.0:
(2008-04-04)
    hooke.py:
        sanity check of CLI plugins to avoid function overloading at startup
    hooke_cli.py ; libhooke.py:
        now playlists keep the index (when you reload the playlist, it starts from the
            last observed curve)
        updated plot to use _send_plot()
        hooke.conf accepts lists as arguments for variables in <display>
        txt, export now have consistent argument order (thanks to A.G.Casado for pointing me that)
        txt crashes no more if no filename is given (thanks to A.G.Casado for pointing me that)
    libhookecurve.py:
        added add_set() , remove_set() methods to make life easier for plugin writers
    procplots.py:
        plotmanip_correct() works with new picoforce.py deflection output (see)
    PLUGINS:
    fit.py:
        updated wlc to use _send_plot()
        wlc noauto now keeps the contact point
        wlc reclick to click again the contact point
        temperature now set in hooke.conf
    generalvclamp.py:
        implemented slope (thanks to Marco Brucale)
        implemented autopeak
    flatfilts.py:
        convfilt,peaks use flattened curve
    macro.py:
        (new) added macro plugin (thanks to Alberto Gomez Casado)
    DRIVERS:
    picoforce.py:
        fixed trigger bug! (thanks to Alberto Gomez Casado)
        better deflection output (separated extension,retraction)

0.7.5:
(2008-03-27)
    hooke_cli.py:
        removed outdated size command
    PLUGINS:
    generalvclamp.py:
        implemented flatten
    DRIVERS:
    added tutorialdriver.py driver
    csvdriver.py:
        fixed (forgot close_all() method)

0.7.4:
(2008-03-19)
    added csvdriver driver
    hooke_cli.py:
        fixed plot manipulators handling (now it's safe to comment a
              plot manipulator on hooke.conf)
    PLUGINS:
    fit.py:
        fixed possible crash when clicking two times the same point on wlc

0.7.3:
(2008-01-10)
    hooke_cli.py:
        fixed crash on copylog
    PLUGINS:
        massanalysis.py:
            Initial release
        tutorial.py:
            Tutorial plugin, initial release

0.7.2.1:
(2007-11-30)
    PLUGINS:
        flatfilt.py:
            fixed crash on Windows

0.7.2:
(2007-11-29)
    hooke.py:
        new configuration variable hookedir
    hooke_cli.py:
        copylog now checks if the destination is a real directory
        fixed crashes in set
    PLUGINS:
        generalvclamp.py:
            fixed a crash in forcebase when picking two times the same point
        flatfilt.py:
            fixed crash due to convfilt.conf impossible to load
            initial implementation of the blind window for convfilt
            initial data set maps (NOT FINISHED)

0.7.1:
(2007-11-26)
    PLUGINS:
        flatfilts.py:
            fixed possible crash in convfilt
            implemented configuration file convfilt.conf
            convfilt defaults are now 5 peaks 5 times more the noise absdev
            implemented convconf
            implemented setconf
        libpeakspot.py:
            fixed:now it really uses noise_absdev

0.7.0:
(2007-11-15)
    hooke_cli.py:
        implemented _send_plot() helper API function
    PLUGINS:
        generalvclamp.py:
            fixed forcebase to work with subtplot
        flatfilts.py:
            implemented convfilt!
            added libpeakspot.py (helping library for convolution filter)

0.6.5:
(2007-11-06)
    hooke_cli.py, hooke.py:
        plateau and contact (unmaintained) deleted and scheduled for re-release in generalvramp
        implemented _measure_N_points()
    PLUGINS:
    generalvclamp.py:
        implemented forcebase
    fit.py:
        wlc now accepts and uses temperature as an argument
        wlc has been cleaned and uses new APIs

0.6.4:
(2007-10-23)
    hooke_cli.py, libhooke.py:
        implemented support for defining order of plotmanip methods in hooke.conf
    hooke_cli.py:
        implemented delta
        implemented point
        attempted fix to bug 0033 (notelog crashing Hooke when using Unicode characters)
    PLUGINS:
    generalvramp.py:
        began to move velocity ramp force spectroscopy-specific things in separate plugin
    procplots.py:
        added detriggerize; "set detrigger" 0/1 disables/enables it.
    DRIVERS:
    picoforce.py:
        removed detriggerize() from driver

0.6.3:
(2007-10-02)
    hooke_cli.py:
        rewritten txt command, now working
    DRIVERS:
    picoforce.py:
        implemented detriggerize() to bypass the Picoforce trigger bug
    PLUGINS:
    superimpose.py:
        implemented plotavgimpose

0.6.2:
(2007-09-27)
    hooke_cli.py:
        fixed error handling in notelog
        smarter handling of directory names in genlist
        unexpected error handling in do_plot()
    hooke.py:

libhooke.py
#!/usr/bin/env python

'''
libhooke.py

General library of internal objects and utilities for Hooke.

Copyright (C) 2006 Massimo Sandal (University of Bologna, Italy).
With algorithms contributed by Francesco Musiani (University of Bologna, Italy)

This program is released under the GNU General Public License version 2.
'''



import libhookecurve as lhc

import scipy
import scipy.signal
import scipy.optimize
import scipy.stats
import numpy
import xml.dom.minidom
import os
import string
import csv

HOOKE_VERSION=['0.8.3_devel', 'Seinei', '2008-04-16']
WX_GOOD=['2.6','2.8'] 
    
class PlaylistXML:
        '''
        This module allows for import/export of an XML playlist into/out of a list of HookeCurve objects
        '''
        
        def __init__(self):
            
            self.playlist=None #the DOM object representing the playlist data structure
            self.playpath=None #the path of the playlist XML file
            self.plaything=None
            self.hidden_attributes=['curve'] #This list contains hidden attributes that we don't want to go into the playlist.
        
        def export(self, list_of_hooke_curves, generics):
            '''
            Creates an initial playlist from a list of files.
            A playlist is an XML document with the following syntaxis:
            <playlist>
            <element path="/my/file/path/"/ attribute="attribute">
            <element path="...">
            </playlist>
            '''   
        
            #create the output playlist, a simple XML document
            impl=xml.dom.minidom.getDOMImplementation()
            #create the document DOM object and the root element
            newdoc=impl.createDocument(None, "playlist",None)
            top_element=newdoc.documentElement
            
            #save generics variables
            playlist_generics=newdoc.createElement("generics")
            top_element.appendChild(playlist_generics)
            for key in generics.keys():
                newdoc.createAttribute(key)
                playlist_generics.setAttribute(key,str(generics[key]))
            
            #save curves and their attributes
            for item in list_of_hooke_curves:
                #playlist_element=newdoc.createElement("curve")
                playlist_element=newdoc.createElement("element")
                top_element.appendChild(playlist_element)
                for key in item.__dict__:
                    if not (key in self.hidden_attributes):
                        newdoc.createAttribute(key)
                        playlist_element.setAttribute(key,str(item.__dict__[key]))    
            
            self.playlist=newdoc
            
        def load(self,filename):
            '''
            loads a playlist file
            '''
            myplay=file(filename)
            self.playpath=filename
            
            #the following 3 lines are needed to strip newlines. otherwise, since newlines
            #are XML elements too (why?), the parser would read them (and re-save them, multiplying
            #newlines...)
            #yes, I'm an XML n00b
            the_file=myplay.read()
            the_file_lines=the_file.split('\n')
            the_file=''.join(the_file_lines)
                       
            self.playlist=xml.dom.minidom.parseString(the_file)  
                           
            #inner parsing functions
            def handlePlaylist(playlist):
                list_of_files=playlist.getElementsByTagName("element")
                generics=playlist.getElementsByTagName("generics")
                return handleFiles(list_of_files), handleGenerics(generics)
            
            def handleGenerics(generics):
                generics_dict={}
                if len(generics)==0:
                    return generics_dict
                
                for attribute in generics[0].attributes.keys():
                    generics_dict[attribute]=generics[0].getAttribute(attribute)
                return generics_dict
        
            def handleFiles(list_of_files):
                new_playlist=[]
                for myfile in list_of_files:
                    #rebuild a data structure from the xml attributes
                    the_curve=lhc.HookeCurve(myfile.getAttribute('path'))
                    for attribute in myfile.attributes.keys(): #extract attributes for the single curve
                        the_curve.__dict__[attribute]=myfile.getAttribute(attribute)
                    new_playlist.append(the_curve)
                
                return new_playlist #this is the true thing returned at the end of this function...(FIXME: clarity)
                    
            return handlePlaylist(self.playlist)
            

        def save(self,output_filename):
            '''
            saves the playlist in a XML file.
            '''
            try:
                outfile=file(output_filename,'w')
            except IOError:
                print 'libhooke.py : Cannot save playlist. Wrong path or filename'
                return
            
            self.playlist.writexml(outfile,indent='\n')
            outfile.close()


class HookeConfig:
    '''
    Handling of Hooke configuration file
    
    Mostly based on the simple-yet-useful examples of the Python Library Reference
    about xml.dom.minidom
    
    FIXME: starting to look a mess, should require refactoring
    '''
    
    def __init__(self):
        self.config={}
        self.config['plugins']=[]
        self.config['drivers']=[]
        self.config['plotmanips']=[]
                
    def load_config(self, filename):
        myconfig=file(filename)
                    
        #the following 3 lines are needed to strip newlines. otherwise, since newlines
        #are XML elements too, the parser would read them (and re-save them, multiplying
        #newlines...)
        #yes, I'm an XML n00b
        the_file=myconfig.read()
        the_file_lines=the_file.split('\n')
        the_file=''.join(the_file_lines)
                       
        self.config_tree=xml.dom.minidom.parseString(the_file)  
        
        def getText(nodelist):
            #take the text from a nodelist
            #from Python Library Reference 13.7.2
            rc = ''
            for node in nodelist:
                if node.nodeType == node.TEXT_NODE:
                    rc += node.data
            return rc
        
        def handleConfig(config):
            display_elements=config.getElementsByTagName("display")
            plugins_elements=config.getElementsByTagName("plugins")
            drivers_elements=config.getElementsByTagName("drivers")
            workdir_elements=config.getElementsByTagName("workdir")
            defaultlist_elements=config.getElementsByTagName("defaultlist")
            plotmanip_elements=config.getElementsByTagName("plotmanips")
            handleDisplay(display_elements)
            handlePlugins(plugins_elements)
            handleDrivers(drivers_elements)
            handleWorkdir(workdir_elements)
            handleDefaultlist(defaultlist_elements)
            handlePlotmanip(plotmanip_elements)
            
        def handleDisplay(display_elements):
            for element in display_elements:
                for attribute in element.attributes.keys():
                    self.config[attribute]=element.getAttribute(attribute)
                    
        def handlePlugins(plugins):
            for plugin in plugins[0].childNodes:
                try:
                    self.config['plugins'].append(str(plugin.tagName))
                except: #if we allow fancy formatting of xml, there is a text node, so tagName fails for it...
                    pass
        #FIXME: code duplication
        def handleDrivers(drivers):
            for driver in drivers[0].childNodes:
                try:
                    self.config['drivers'].append(str(driver.tagName))
                except: #if we allow fancy formatting of xml, there is a text node, so tagName fails for it...
                    pass
        
        def handlePlotmanip(plotmanips):
            for plotmanip in plotmanips[0].childNodes:
                try:
                    self.config['plotmanips'].append(str(plotmanip.tagName))
                except: #if we allow fancy formatting of xml, there is a text node, so tagName fails for it...
                    pass
        
        def handleWorkdir(workdir):
            '''
            default working directory
            '''
            wdir=getText(workdir[0].childNodes)
            self.config['workdir']=wdir.strip()
            
        def handleDefaultlist(defaultlist):
            '''
            default playlist
            '''
            dflist=getText(defaultlist[0].childNodes)
            self.config['defaultlist']=dflist.strip()
            
        handleConfig(self.config_tree)
        #making items in the dictionary more machine-readable
        for item in self.config.keys():
            try:
                self.config[item]=float(self.config[item])
            except TypeError: #we are dealing with a list, probably. keep it this way.
                try:
                    self.config[item]=eval(self.config[item])
                except: #not a list, not a tuple, probably a string?
                    pass
            except ValueError: #if we can't get it to a number, it must be None or a string
                if string.lower(self.config[item])=='none':
                    self.config[item]=None
                else:
                    pass
                                                
        return self.config
        
        
    def save_config(self, config_filename):
        print 'Not Implemented.'
        pass    


class ClickedPoint:
    '''
    this class defines what a clicked point on the curve plot is
    '''
    def __init__(self):
        
        self.is_marker=None #boolean ; decides if it is a marker
        self.is_line_edge=None #boolean ; decides if it is the edge of a line (unused)
        self.absolute_coords=(None,None) #(float,float) ; the absolute coordinates of the clicked point on the graph
        self.graph_coords=(None,None) #(float,float) ; the coordinates of the plot that are nearest in X to the clicked point
        self.index=None #integer ; the index of the clicked point with respect to the vector selected
        self.dest=None #0 or 1 ; 0=top plot 1=bottom plot
                
        
    def find_graph_coords_old(self, xvector, yvector):
        '''
        Given a clicked point on the plot, finds the nearest point in the dataset (in X) that
        corresponds to the clicked point.
        OLD & DEPRECATED - to be removed
        '''
                   
        #FIXME: a general algorithm using min() is needed!
        print '---DEPRECATED FIND_GRAPH_COORDS_OLD---'
        best_index=0
        best_dist=10**9 #should be more than enough given the scale
                
        for index in scipy.arange(1,len(xvector),1):
            dist=((self.absolute_coords[0]-xvector[index])**2)+(100*((self.absolute_coords[1]-yvector[index])))**2
                        #TODO, generalize? y coordinate is multiplied by 100 due to scale differences in the plot
            if dist<best_dist:
                best_index=index
                best_dist=dist
                        
        self.index=best_index
        self.graph_coords=(xvector[best_index],yvector[best_index])
        return
            
    def find_graph_coords(self,xvector,yvector):
        '''
        Given a clicked point on the plot, finds the nearest point in the dataset that
        corresponds to the clicked point.
        '''
        dists=[]
        for index in scipy.arange(1,len(xvector),1):
            dists.append(((self.absolute_coords[0]-xvector[index])**2)+((self.absolute_coords[1]-yvector[index])**2))
                        
        self.index=dists.index(min(dists))
        self.graph_coords=(xvector[self.index],yvector[self.index])
#-----------------------------------------
#CSV-HELPING FUNCTIONS        
        
def transposed2(lists, defval=0):
    '''
    transposes a list of lists, i.e. from [[a,b,c],[x,y,z]] to [[a,x],[b,y],[c,z]] without losing
    elements
    (by Zoran Isailovski on the Python Cookbook online)
    '''
    if not lists: return []
    return map(lambda *row: [elem or defval for elem in row], *lists)
        
def csv_write_dictionary(f, data, sorting='COLUMNS'):
    '''
    Writes a CSV file from a dictionary, with keys as first column or row
    Keys are in "random" order.
    
    Keys should be strings
    Values should be lists or other iterables
    '''
    keys=data.keys()
    values=data.values()
    t_values=transposed2(values)
    writer=csv.writer(f)

    if sorting=='COLUMNS':
        writer.writerow(keys)
        for item in t_values:
            writer.writerow(item)
        
    if sorting=='ROWS':
        print 'Not implemented!' #FIXME: implement it.


#-----------------------------------------        
                    
def debug():
    '''
    debug stuff from latest rewrite of hooke_playlist.py
    should be removed sooner or later (or substituted with new debug code!)
    '''
    confo=HookeConfig()
    print confo.load_config('hooke.conf')

if __name__ == '__main__':
    debug()